packages/skill-lesson/hooks/use-progress.ts

```diff
@@ -3,31 +3,165 @@ import {SanityDocument} from '@sanity/client'
+import {Exercise} from '../schemas/exercise'
+import {Section} from '../schemas/section'
+import {Module} from '../schemas/module'
+
+export function extractExercisesAndResource(sections?: Section[] | null) {
+  const exercises = flatMap<Section, Exercise>(
+    sections,
+    (section) => section?.lessons || [],
+  )
+  const resources = flatMap<Section, Lesson>(sections, (section) =>
+    map(section.lessons, (lesson: {solution?: any}) =>
+      lesson.solution ? lesson.solution : lesson,
+    ),
+  )
 
-export const useModuleProgress = ({module}: {module: SanityDocument}) => {
+  return {exercises, resources}
+}
+
+export const getLastCompletedSolution = ({
+  resources,
+  moduleProgress,
+}: {
+  resources: Lesson[]
+  moduleProgress: LessonProgress[]
+}) => {
+  return !isEmpty(moduleProgress)
+    ? resources.find((resource) => {
+        return resource.slug === get(first(moduleProgress), 'lessonSlug')
+      })
+    : first(resources)
+}
+
+export const getNextSectionAndModuleProgress = ({
+  userProgress,
+  resources,
+  lastCompletedExercise,
+  sections,
+  module,
+}: {
+  sections?: Section[] | null
+  userProgress: LessonProgress[]
+  resources: Lesson[]
+  lastCompletedExercise?: Lesson | null
+  module: Module
+}) => {
+  const moduleProgress = userProgress.filter(({completedAt, lessonSlug}) => {
+    if (!completedAt) return
+    return find(resources, {slug: lessonSlug})
+  })
+
+  const activeSection =
+    first(
+      sections?.filter(({lessons}) => {
+        const lesson = find(lessons, {slug: lastCompletedExercise?.slug})
+        return lesson
+      }),
+    ) || null
+
+  const isLastLessonInSection = activeSection?.lessons
+    ? indexOf(activeSection?.lessons, lastCompletedExercise) ===
+      activeSection?.lessons.length - 1
+    : false
+
+  const nextSection =
+    activeSection && isLastLessonInSection
+      ? getNextSection({
+          module: module,
+          currentSection: activeSection,
+        })
+      : activeSection
+
+  return {nextSection, moduleProgress}
+}
+
+export const getCompleted = ({
+  resources,
+  moduleProgress,
+  exercises,
+  nextExercise,
+}: {
+  resources: Lesson[]
+  moduleProgress: LessonProgress[]
+  exercises: Exercise[]
+  nextExercise?: Lesson | null
+}) => {
+  const isCompleted = moduleProgress.length === resources.length
+  const percentCompleted = Math.round(
+    (moduleProgress.length / resources.length) * 100,
+  )
+
+  const hasProgress = !isEmpty(
+    find(moduleProgress, {
+      // if explainer
+      lessonSlug: get(find(exercises, {slug: nextExercise?.slug}), 'slug'),
+    }) ||
+      find(moduleProgress, {
+        lessonSlug: get(
+          find(exercises, {slug: nextExercise?.slug}),
+          'solution.slug',
+        ),
+      }),
+  )
+
+  const lastCompletedSolution = getLastCompletedSolution({
+    resources,
+    moduleProgress,
+  })
+
+  return {isCompleted, percentCompleted, lastCompletedSolution, hasProgress}
+}
```

Comment:
extracted these methods for testing seams

https://refactoring.guru/extract-method